File: fsharp/005/005.fs
let prime_factors n =
  let rec pf current possible_factor =
    if current < (possible_factor * possible_factor) then [current]
    elif current % possible_factor = 0I then possible_factor :: (pf (current / possible_factor) possible_factor)
    else pf current (possible_factor + 1I)
  pf n 2I

let rec remove_first i l =
  match l with
  | head :: tail when head = i -> tail
  | head :: tail -> head :: remove_first i tail
  | _ -> []

let overlap a b =
  let rec o a b =
    match a, b with
    | head :: tail, b -> o tail (remove_first head b)
    | a, [] -> a
    | [], b -> b
  List.append a (o a b) |> List.sort

printf "%A\n" ([1I..20I] |> List.map (fun x -> prime_factors x) |> List.reduce (fun a b -> overlap a b) |> List.reduce (fun a b -> a * b))

File: scala/005/005.scala
object Euler {
  def main(args: Array[String]): Unit = {
    println((1 until 21).map(x => primeFactors(x)).reduceLeft((x, y) => overlap(x, y)).reduceLeft((x, y) => x * y))
  }

  def primeFactors(n: Long): List[Long] = {
    def pf(tuple: (Long, Long)): List[Long] = {
      tuple match {
        case (n, p) if (n < p * p)  => List[Long](n)
        case (n, p) if (n % p == 0) => List[Long](p) ++ pf((n / p, p))
        case (n, p)                 => pf((n, p + 1))
      }
    }

    pf((n, 2))
  }

  def removeFirst[T](list: List[T])(pred: (T) => Boolean): List[T] = {
    val (before, after) = list span (x => !pred(x))

    before ::: after.drop(1)
  }

  def overlap[T](a: List[T], b: List[T]): List[T] = {
    def o[T](x: List[T], y: List[T]): List[T] = {
      if (x.length == 0) {
        y
      } else {
        o(x.tail, removeFirst(y) { _ == x.head })
      }
    }

    a ++ o(a, b)
  }
}
